[HOLD] Blazor WASM debug topic updates for 3.2p3

guardrex opened the pull request:
Fixes #17192

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/debug?view=aspnetcore-3.1&branch=pr-en-us-17438)

guardrex commented:
@danroth27 This is updated now from your content.

danroth27 commented on aspnetcore/blazor/debug.md:
```diff
@@ -15,28 +15,27 @@ uid: blazor/debug
 [!INCLUDE[](~/includes/blazorwasm-preview-notice.md)]
 
-*Early* support exists for debugging Blazor WebAssembly using the browser dev tools in Chromium-based browsers (Chrome/Edge). Work is in progress to:
+As of ASP.NET Core 3.2 Preview 3, Blazor WebAssembly apps can be debugged using the browser dev tools in Chromium-based browsers (Chrome/Edge) in Visual Studio and Visual Studio Code.
```
```suggestion
Blazor WebAssembly apps can be debugged using the browser dev tools in Chromium-based browsers (Edge/Chrome).  Alternatively you can debug your app using Visual Studio or Visual Studio Code.
```

danroth27 commented on aspnetcore/blazor/debug.md:
```diff
@@ -45,54 +44,117 @@ Debugging requires either of the following browsers:
 
+## Enable debugging
+
+To enable debugging in an existing Blazor WebAssembly app, update *launchSettings.json* in the startup project of your app. Set `useWebAssemlbyDebugging` to `true` and include the following `inspectUri` property in each launch profile:
```
Turns out `useWebAssemblyDebugging` isn't necessary and shouldn't be in the template at all.

```suggestion
To enable debugging in an existing Blazor WebAssembly app in Visual Studio or Visual Studio Code, update *launchSettings.json* in the startup project of your app to include the following `inspectUri` property in each launch profile:
```

danroth27 commented on aspnetcore/blazor/debug.md:
````diff
@@ -45,54 +44,117 @@ Debugging requires either of the following browsers:
+ 
+```json
+{
+  "useWebAssemblyDebugging": true,
````
The `useWebAssemblyDebugging` property actually isn't needed. It still shows up in our templates, but we'll get that cleaned up in the next release.

```suggestion
```

danroth27 commented on aspnetcore/blazor/debug.md:
```diff
@@ -45,54 +44,117 @@ Debugging requires either of the following browsers:
+The preceding properties:
+
+* Enable the IDE to detect that the app is a Blazor WebAssembly app.
+* Instruct the script debugging infrastructure to connect to the browser through Blazor's debugging proxy.
```
```suggestion
The `inspectUri property:

* Enables the IDE to detect that the app is a Blazor WebAssembly app.
* Instructs the script debugging infrastructure to connect to the browser through Blazor's debugging proxy.
```

danroth27 commented on aspnetcore/blazor/debug.md:
```diff
@@ -102,23 +164,10 @@ If remote debugging is disabled, an **Unable to find debuggable browser tab** er
 ## Debug the app
 
-Once Chrome is running with remote debugging enabled, the debugging keyboard shortcut opens a new debugger tab. After a moment, the **Sources** tab shows a list of the .NET assemblies in the app. Expand each assembly and find the *.cs*/*.razor* source files available for debugging. Set breakpoints, switch back to the app's tab, and the breakpoints are hit when the code executes. After a breakpoin ...
+Once Chrome is running with remote debugging enabled, the debugging keyboard shortcut opens a new debugger tab. After a moment, the **Sources** tab shows a list of the .NET assemblies in the app. Expand each assembly and find the *.cs*/*.razor* source files available for debugging. Set breakpoints, switch back to the app's tab, and the breakpoints are hit when the code executes. After a breakpoin ...
```
```suggestion
Once the browser is running with remote debugging enabled, the debugging keyboard shortcut opens a new debugger tab. After a moment, the **Sources** tab shows a list of the .NET assemblies in the app. Expand each assembly and find the *.cs*/*.razor* source files available for debugging. Set breakpoints, switch back to the app's tab, and the breakpoints are hit when the code executes. After a breakpoint is hit, single-step (<kbd>F10</kbd>) through the code or resume (<kbd>F8</kbd>) code execution normally.
```

danroth27 commented on aspnetcore/blazor/debug.md:
```diff
@@ -45,54 +44,117 @@ Debugging requires either of the following browsers:
 * Google Chrome (version 70 or later)
 * Microsoft Edge preview ([Edge Dev Channel](https://www.microsoftedgeinsider.com))
```
The Chromium based Microsoft Edge is no longer in preview:  https://microsoft.com/edge. If you have Edge version 80 or later, you should be good to go.

danroth27 commented on aspnetcore/blazor/debug.md:
```diff
@@ -45,54 +44,117 @@ Debugging requires either of the following browsers:
 * Google Chrome (version 70 or later)
```
Let's switch the order and put Edge first.

danroth27 commented on aspnetcore/blazor/debug.md:
```diff
@@ -15,28 +15,27 @@ uid: blazor/debug
-* Single-step (`F10`) through the code or resume (`F8`) code execution.
-* In the *Locals* display, observe the values of any local variables of type `int`, `string`, and `bool`.
+* Run the app with debugging support in Visual Studio (<kbd>F5</kbd> support).
+* Single-step (<kbd>F10</kbd>) through the code or resume (<kbd>F8</kbd>) code execution.
```
F8 is only used for resuming execution in a browser. In VS or VS Code you'd use F5.

danroth27 commented on aspnetcore/blazor/debug.md:
```diff
@@ -45,54 +44,117 @@ Debugging requires either of the following browsers:
 * Google Chrome (version 70 or later)
 * Microsoft Edge preview ([Edge Dev Channel](https://www.microsoftedgeinsider.com))
 
+## Enable debugging
```
Adding `inspectUri` is only needed for debugging in our IDEs. If you just want to debug in the browser, then this isn't really required.

In addition to debugging in our IDEs, we still need cover how to debug in a browser using the browser dev tools.
1. Run a Debug build of the app in the Development environment
1. Hit <kbd>Shift-Alt-D</kbd>
1. Follow the instructions to restart the browser with remote debugging enabled if needed.
1. Use the browser dev tools to debug your app

Then we can cover debugging in an IDE

guardrex commented:
@danroth27 Updates made.

* We don't have approval for a tab (i.e., "Debug in the browser) to go along with the VS and VSC tabs, so I split all three out into sections.

* I changed the title to include "WebAssembly."

* For ...

  * Run the app with debugging support in Visual Studio and Visual Studio Code (<kbd>F5</kbd> support).
  * Single-step (<kbd>F10</kbd>) through the code.
  * Resume code execution in a browser with <kbd>F8</kbd> or <kbd>F5</kbd> in Visual Studio or Visual Studio Code.

  ... does it need a bit more work?

danroth27 commented on aspnetcore/blazor/debug.md:
```diff
@@ -1,123 +1,180 @@
+* Run the app with debugging support in Visual Studio and Visual Studio Code (<kbd>F5</kbd> support).
+* Single-step (<kbd>F10</kbd>) through the code.
+* Resume code execution in a browser with <kbd>F8</kbd> or <kbd>F5</kbd> in Visual Studio or Visual Studio Code.
+* In the *Locals* display, observe the values of any local variables.
```
```suggestion
* In the *Locals* display, observe the values of local variables.
```

I think we still have some issues viewing certain value types.

danroth27 commented on aspnetcore/blazor/debug.md:
```diff
@@ -1,123 +1,180 @@
+
+## Enable remote debugging
+
+If remote debugging is disabled, an **Unable to find debuggable browser tab** error page is generated by Chrome. The error page contains instructions for running Chrome with the debugging port open so that the Blazor debugging proxy can connect to the app. *Close all Chrome instances* and restart Chrome as instructed.
```
```suggestion
To debug a Blazor WebAssembly app the browser must be run with remote debugging enabled. If remote debugging is disabled, an **Unable to find debuggable browser tab** error page is generated. The error page contains instructions for running the browser with the debugging port open so that the Blazor debugging proxy can connect to the app. *Close all browser instances* and restart the browser as instructed.
```

danroth27 commented on aspnetcore/blazor/debug.md:
```diff
@@ -1,123 +1,180 @@
+* Enables the IDE to detect that the app is a Blazor WebAssembly app.
+* Instructs the script debugging infrastructure to connect to the browser through Blazor's debugging proxy.
 
 ## Procedure
```
Instead of having "Procedure" as section that shows up in the right nav, can we remove the Procedure heading and elevate the following sections to be h2s?

danroth27 commented on aspnetcore/blazor/debug.md:
```diff
@@ -1,123 +1,180 @@
+   ![Debug Counter in VS Code](https://devblogs.microsoft.com/aspnet/wp-content/uploads/sites/16/2020/03/vscode-debug-counter.png)
 
-Attaching to the app's process in Visual Studio is a *temporary* debugging scenario for Blazor WebAssembly while **F5** debugging is in development.
+### Debug in the browser
```
It loos like the "Enable remote debugging" and "Debug the app" sections below should just be merged into this section as both are about in-browser debugging.

mkArtakMSFT commented on aspnetcore/blazor/debug.md:
```diff
@@ -1,123 +1,170 @@
+* Debug an app with a service worker.
 
-Development of further debugging scenarios is an on-going focus of the engineering team.
+We expect to continue to improve the debugging experience in future releases.
```
My own perception of "future" is a bit vague and it leaves me with an impression that this may not happen.
What about slightly modified version below (please ignore this, if you find your phrasing better):

```suggestion
We will continue to improve the debugging experience in upcoming releases.
```

mkArtakMSFT commented on aspnetcore/blazor/debug.md:
```diff
@@ -1,123 +1,170 @@
-* Microsoft Edge preview ([Edge Dev Channel](https://www.microsoftedgeinsider.com))
 
-## Procedure
+## Enable debugging for Visual Studio and Visual Studio Code
```
Should this rather be 
```suggestion
## Enable debugging for existing Blazor WebAssembly projects

Debugging should work out of the box for projects created using 3.2-preview3 and later project templates.
For existing projects, , however, a gesture is required.
```

guardrex replied:
No, it should say that explicitly because the following instruction doesn't apply to debugging in the browser. The real problem here is that we can't have **_custom tabs_**. If we could have custom tabs, then this content could go **only** at the tops of the VS and VSC tabs. The 'Debug in the browser' tab wouldn't get it.

mkArtakMSFT replied:
Ah, good point! I think the title can remain in that case. You're choice whether the other sentence is worthy or not

guardrex replied:
Yes ... I'll fix that now. My :eye: went straight to the highlighted text in the diff, so I didn't pick up on the rest. I'll fix it now.

guardrex commented:
@mkArtakMSFT See :eyes: how this looks :point_right: https://github.com/dotnet/AspNetCore.Docs/pull/17438/commits/ac0ceddec9f4d3f9f117ccbbcad24bd98a720e60

guardrex commented:
I'll patch it if it needs more work. Going ahead with the merge now, and I'll take this batch live immediately.
